src/main/scala/com/databricks/spark/avro/DefaultSource.scala

```diff
@@ -15,20 +15,67 @@
  */
 package com.databricks.spark.avro
 
-import org.apache.spark.sql.SQLContext
-import org.apache.spark.sql.sources.RelationProvider
+import org.apache.hadoop.fs.Path
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, SaveMode, SQLContext}
+import org.apache.spark.sql.sources._
 
 /**
  * Provides access to Avro data from pure SQL statements (i.e. for users of the
  * JDBC server).
  */
-class DefaultSource extends RelationProvider {
+class DefaultSource
+  extends RelationProvider with SchemaRelationProvider with CreatableRelationProvider {
+
+  private def checkPath(parameters: Map[String, String]): String = {
+    parameters.getOrElse("path", sys.error("'path' must be specified for AVRO data."))
+  }
 
   /**
    * Creates a new relation for data store in avro given a `path` as a parameter.
    */
-  def createRelation(sqlContext: SQLContext, parameters: Map[String, String]) = {
-    AvroRelation(parameters("path"))(sqlContext)
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String]): BaseRelation = {
+    val minPartitions = parameters.getOrElse("minPartitions", "0").toInt
+    AvroRelation(checkPath(parameters), None, minPartitions)(sqlContext)
   }
-}
 
+  override def createRelation(
+      sqlContext: SQLContext,
+      parameters: Map[String, String],
+      schema: StructType): BaseRelation = {
+    val minPartitions = parameters.getOrElse("minPartitions", "0").toInt
+    AvroRelation(checkPath(parameters), Some(schema), minPartitions)(sqlContext)
+  }
+
+  override def createRelation(
+      sqlContext: SQLContext,
+      mode: SaveMode,
+      parameters: Map[String, String],
+      data: DataFrame): BaseRelation = {
+    val path = parameters("path")
+    val filesystemPath = new Path(path)
+    val fs = filesystemPath.getFileSystem(sqlContext.sparkContext.hadoopConfiguration)
+    val doSave = if (fs.exists(filesystemPath)) {
+      mode match {
+        case SaveMode.Append =>
+          sys.error(s"Append mode is not supported by ${this.getClass.getCanonicalName}")
```

Comment:
getName is a little more robust, not returning null for anonymous inner classes (not that this is one, but I think getName is typically preferred -- canonical is apparently how you would import, name is how you can Class.forName())